### PR TITLE
#2 Add disk create to VM resource

### PR DESCRIPTION
**Pull request by ja6a**
Add new disk element to the VM resource. 
The allows the ability to add disks to the VM.
It is a Force New at this stage.
Refactor to make separation between create and update flows cleaner.

Note: rebased on feature/cpu because it is not possible to merge to master at this time.

### CODE REVIEW

**Comment by pegerto** [comment]
Requires rebase

**Review by pegerto — CHANGES_REQUESTED** [review]
Rebase necesary

File: skytap/resource_skytap_vm.go [hunk 1]
```diff
@@ -179,8 +217,43 @@ func resourceSkytapVMCreate(d *schema.ResourceData, meta interface{}) error {
 	if err = addNetworkAdapters(d, meta, *vm.ID); err != nil {
 		return err
 	}
+	if err = addVMHardware(d, &client, &ctx, vm, environmentID, vmID); err != nil {
+		return err
+	}
+
+	forceRunning := true
```
Comment (pegerto) [inline skytap/resource_skytap_vm.go:224]: We should not change the behaviour of the code during testing.

File: skytap/resource_skytap_vm.go [hunk 2]
```diff
@@ -337,48 +410,65 @@ func resourceSkytapVMRead(d *schema.ResourceData, meta interface{}) error {
...
-	d.Set("environment_id", environmentID)
-	d.Set("name", vm.Name)
-	d.Set("cpus", vm.Hardware.CPUs)
-	d.Set("name", vm.Hardware.RAM)
+	err = d.Set("environment_id", environmentID)
+	if err != nil {
+		return err
+	}
+	err = d.Set("name", vm.Name)
+	if err != nil {
```
Comment (pegerto) [inline skytap/resource_skytap_vm.go:418]: d.Set of vm.Name been vm.Name a single value should be safe.

File: skytap/resource_skytap_vm.go [hunk 3]
```diff
@@ -337,48 +410,65 @@ func resourceSkytapVMRead(d *schema.ResourceData, meta interface{}) error {
 
 	// templateID and vmID are not set, as they are not returned by the VM response.
 	// If any of these attributes are changed, this VM will be rebuilt.
-	d.Set("environment_id", environmentID)
-	d.Set("name", vm.Name)
-	d.Set("cpus", vm.Hardware.CPUs)
-	d.Set("name", vm.Hardware.RAM)
+	err = d.Set("environment_id", environmentID)
```
Comment (pegerto) [inline skytap/resource_skytap_vm.go:413]: d.Set of int should be safe.

**Review by pegerto — CHANGES_REQUESTED** [review]
Missing an acceptance test.

File: skytap/resource_skytap_vm.go [hunk 4]
```diff
@@ -455,23 +540,28 @@ func updateVMResourceOptions(d *schema.ResourceData, client *skytap.VMsService,
...
 		opts.Hardware.RAM = utils.Int(v.(int))
 	}
+	if v, ok := d.GetOk("disk"); ok {
+		if opts.Hardware == nil {
+			opts.Hardware = &skytap.UpdateHardware{}
+		}
+		opts.Hardware.NewDisks = &skytap.NewDisks{}
+		disks := v.(*schema.Set)
+		log.Printf("[INFO] creating %d disks", disks.Len())
+		opts.Hardware.NewDisks.Sizes = make([]int, d.Get("disk.#").(int))
```
Comment (pegerto) [inline skytap/resource_skytap_vm.go:550]: Is there a max number of disk that we can create?

Comment (ja6a) [reply skytap/resource_skytap_vm.go:550]: This is not specified in the docs AFAIK.

File: skytap/resource_skytap_vm.go (hunk 3 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:413]: I propose to remove the nil checks in a future PR. I will update this comment when this is done.

File: skytap/resource_skytap_vm.go (hunk 1 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:224]: I propose to remove this in a future PR. I will update this comment when this is done.

File: skytap/resource_skytap_vm.go (hunk 2 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:418]: I propose to remove the nil checks in a future PR. I will update this comment when this is done.

**Comment by ja6a** [comment]
The acceptance test is added in the next PR.

**Review by pegerto — APPROVED** [review]

File: skytap/resource_skytap_vm.go (hunk 1 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:224]: Removed

File: skytap/resource_skytap_vm.go (hunk 3 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:413]: error checks removed.

File: skytap/resource_skytap_vm.go (hunk 2 above)
Comment (ja6a) [reply skytap/resource_skytap_vm.go:418]: error checks removed.

**Pull request merged**